src/Stack/Setup.hs

```diff
@@ -669,6 +674,91 @@ ensureCompiler sopts = do
       , cpGlobalDB' = globaldb
       }
 
+buildGhcFromSource :: forall env.
+   ( HasTerm env
+   , HasProcessContext env
+   , HasConfig env
+   ) => Memoized SetupInfo -> [Tool] -> CompilerRepository -> Text -> Text
+   -> RIO env (Either (ActualCompiler, Arch, Path Abs Dir) Tool, CompilerBuild)
+buildGhcFromSource getSetupInfo' installed (CompilerRepository url) commitId flavour = do
+   config <- view configL
+   let compilerTool = ToolGhcGit commitId flavour
+
+   -- detect when the correct GHC is already installed
+   if compilerTool `elem` installed
+     then return (Right compilerTool,CompilerBuildStandard)
+     else do
+       let repo = Repo
+            { repoCommit = commitId
+            , repoUrl    = url
+            , repoType   = RepoGit
+            , repoSubdir = mempty
+            }
+
+       -- clone the repository and execute the given commands
+       Pantry.withRepo repo $ do
+         threads <- view $ configL.to configJobs
+         let
+           hadrianArgs = fmap T.unpack
+               [ hadrianCmd
+               , "-c"                    -- run ./boot and ./configure
+               , "-j" <> tshow threads   -- parallel build
+               , "--flavour=" <> flavour -- selected flavour
+               , "binary-dist"
+               ]
+           hadrianCmd
+             | osIsWindows = "./hadrian/build.stack.bat"
+             | otherwise   = "./hadrian/build.stack.sh"
+
+         logSticky $ "Building GHC from source with `"
+            <> RIO.display flavour
+            <> "` flavour. It can take a long time (more than one hour)..."
+
+         -- RIO.Process doesn't wrap process' "shell".
+         -- Instead we use "proc" with the "sh" command
+         proc "sh" hadrianArgs runProcess_
```

Comment:
I don't think this will work on Windows, since we're trying to run a batch file, not a shell script. What about instead:

```haskell
proc hadrianCmd hadrianArgs runProcess_
```

And removing the `hadrianCmd` from the args above.

Reply:
IIRC it wasn't working this way on Linux: we would need [shell](https://www.stackage.org/haddock/lts-13.16/process-1.6.3.0/System-Process.html#v:shell) but RIO only has [proc](https://www.stackage.org/haddock/lts-13.16/rio-0.1.8.0/RIO-Process.html#v:proc) AFAIK.

Reply:
Is the shell script in the repo set to be executable?

Reply:
Yes it is. It fails with "Did not find executable at specified path: ./hadrian/build.stack.sh"

Reply:
Please see 2768bb9b1fa3a73bf2a4903d583b526d64e2e78c. This is a really confusing aspect of the process package, I only figured this out since I've been bitten by it before.